src/util/src/WaveformUtil.cc

```diff
@@ -0,0 +1,251 @@
+    Log::Die("WaveformUtil: Start of lookback window must be before peak.");
+  } else if (peakSample > waveform.size()) {
+    Log::Die("WaveformUtil: Peak must be before end of waveform.");
+  }
+
+  // Start at the peak and scan backwards
+  for (int i = peakSample; i > back_window; i--) {
+    double voltage = waveform[i];
+
+    if (voltage > voltageThreshold) {
+      thresholdCrossing = i;
+      break;
+    }
+
+    // Reached the begining of the waveform
+    // returned an invalid value
+    if (i == back_window) {
+      thresholdCrossing = INVALID;  // Invalid value for bad waveforms
+      break;
+    }
+  }
+  return thresholdCrossing;
+}
+
+int GetNCrossings(const std::vector<double>& waveform, double voltageThreshold) {
+  /*
+  Calculates the total number of threshold crossings
+  */
+  int nCrossings = 0;
+
+  bool crossed = false;
+  // Scan over the entire waveform
+  for (int i = 0; i < waveform.size(); i++) {
+    double voltage = waveform[i];
+
+    // If we crossed below threshold
+    if (voltage < voltageThreshold) {
+      // Not already below thresh, count the crossing
+      if (!crossed) {
+        nCrossings += 1;
+      }
+      // Mark that we crossed
+      crossed = true;
+    }
+    // If we are above threshold
+    if (voltage > voltageThreshold) {
+      crossed = false;
+    }
+  }
+  return nCrossings;
+}
+
+std::tuple<int, double, double> GetCrossingsInfo(const std::vector<double>& waveform, double voltageThreshold,
+                                                 double timeStep) {
+  /*
+  Calculates the total number of threshold crossings, time over threshold, and voltage over threshold
+  */
+  int nCrossings = 0;
+  double timeOverThreshold = 0;
+  double voltageOverThreshold = 0;
+
+  bool crossed = false;
+  // Scan over the entire waveform
+  for (int i = 0; i < waveform.size(); i++) {
+    double voltage = waveform[i];
+
+    // If we crossed below threshold
+    if (voltage < voltageThreshold) {
+      // Not already below thresh, count the crossing
+      if (!crossed) {
+        nCrossings += 1;
+      }
+      // Count the time over threshold, mark that we crossed
+      timeOverThreshold += timeStep;
+      voltageOverThreshold += voltage;
+      crossed = true;
+    }
+    // If we are above threshold
+    if (voltage > voltageThreshold) {
+      crossed = false;
+    }
+  }
+  return std::make_tuple(nCrossings, timeOverThreshold, voltageOverThreshold);
+}
+
+double CalculateTimeCFD(const std::vector<double>& waveform, std::pair<int, double> peak, double constFrac,
+                        int lookBack, double timeStep) {
+  /*
+  Apply constant-fraction discriminator for a given peak
+  */
+  double voltageThreshold = constFrac * peak.second;
+  int time = GetThresholdCrossingBeforePeak(waveform, peak.first, voltageThreshold, lookBack, timeStep);
+  // Linearly interpolate threshold crossing time
+  double deltav = waveform.at(time + 1) - waveform.at(time);
```

Comment:
ok -- I think we should just treat this stupid edge case. If `time` is the last element in the vector, don't do the interpolation.